x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java

```diff
@@ -421,6 +422,22 @@
     // can handle.
     public static final Setting<Integer> MAX_MACHINE_MEMORY_PERCENT =
             Setting.intSetting("xpack.ml.max_machine_memory_percent", 30, 5, 200, Property.Dynamic, Property.NodeScope);
+    /**
+     * This boolean value indicates if `max_machine_memory_percent` should be ignored and a automatic calculation is used instead.
+     *
+     * This calculation takes into account total node size and the size of the JVM on that node.
+     *
+     * If the calculation fails, we fall back to `max_machine_memory_percent`.
+     *
+     * This setting is NOT dynamic. This allows the cluster administrator to set it on startup without worry of it
+     * being edited accidentally later.
+     * Consequently, it could be that this setting differs between nodes. But, we only ever pay attention to the value
+     * that is set on the current master. As master nodes are responsible for persistent task assignments.
+     */
+    public static final Setting<Boolean> USE_AUTO_MACHINE_MEMORY_PERCENT = Setting.boolSetting(
+        "xpack.ml.use_auto_machine_memory_percent",
+        false,
+        Property.NodeScope);
```

Comment:
The reasons for adding a new setting instead of expanding `xpack.ml.max_machine_memory_percent` to allow a special `-1` value are:

- An old node joining the cluster would not be able to deserialize the setting and would throw errors
- When calculating the memory possible on an old node, we would fail to gather the JVM attribute (as it wouldn't be available). So, we need a sane fallback. Which is the other static calculated value `xpack.ml.max_machine_memory_percent`

Defaulting to `false` keeps users from having things change unexpectedly during a rolling upgrade. 

In the scenario where the administrator does not want to bother setting the max memory percent, they can simply set this value `true`. The assumption is that the ml nodes are focused to only run ML.